riot/riot.py

```diff
@@ -62,60 +66,157 @@ def to_list(x: t.Union[_K, t.List[_K]]) -> t.List[_K]:
+        >>> Interpreter._matches(Version("3.8.2"), Version("3.6"))
+        False
+        >>> Interpreter._matches(Version("3.8.2"), Version("3.7"))
+        False
+        >>> Interpreter._matches(Version("3.8.2"), Version("2"))
+        False
+        >>> Interpreter._matches(Version("3.8.2-dev"), Version("3.8.2"))
+        True
+        >>> Interpreter._matches(Version("3.8.2-alpha"), Version("3.8.2"))
+        True
+
+        FIXME: unfortunately Version will set undefined segments to 0 so
+        we cannot distinguish between "3.0.0" and "3".
+        >>> Interpreter._matches(Version("3.8.2"), Version("3.0.0"))
+        True
+        >>> Interpreter._matches(Version("3.8.2"), Version("3.8.0"))
+        True
+        """
+        if v2 == v1:
+            return True
+        elif v2.micro != v1.micro and v2.micro != 0:
+            return False
+        elif v2.minor != v1.minor and v2.minor != 0:
+            return False
+        elif v2.major != v1.major and v2.major != 0:
+            return False
+        return True
+
+    @classmethod
     @functools.lru_cache()
-    def version(self) -> str:
-        path = self.path()
+    def find(cls) -> t.List["Interpreter"]:
+        """Find all Python interpreters discoverable on the PATH."""
+        ex_dirs = os.environ.get("PATH", "").split(":")
+        # Use a list because order matters (need the first match).
+        interp_paths: t.List[str] = []
+        # packaging says that the regular expression needs to be compiled
+        # with the following flags
+        expr = re.compile(f"python{VERSION_PATTERN}$", re.IGNORECASE | re.VERBOSE)
+        for d in ex_dirs:
+            if not os.path.isdir(d):
+                continue
+            for f in os.listdir(d):
+                if expr.match(f):
+                    interp_paths.append(os.path.join(d, f))
 
-        # Redirect stderr to stdout because Python 2 prints version on stderr
-        output = subprocess.check_output([path, "--version"], stderr=subprocess.STDOUT)
-        version = output.decode().strip().split(" ")[1]
-        return version
+        interps: t.List[Interpreter] = []
+        for i in interp_paths:
+            try:
+                interp = cls(i)
+            except InvalidVersion:
+                logger.warning(
+                    "Failed to parse version for interpreter %r.", i, exc_info=True
+                )
+            else:
+                interps.append(interp)
 
-    @functools.lru_cache()
-    def path(self) -> str:
-        """Return the Python interpreter path or raise.
+        return interps
+
+    @classmethod
+    def find_match(cls, v: Version) -> "Interpreter":
+        """Return a matching interpreter for the given version.
+
+        If one is not found then raises a FileNotFoundError.
+        """
+        interps = [i for i in cls.find() if i.matches(v)]
+        if len(interps):
+            return interps[0]
+        raise FileNotFoundError("No interpreter matching %r", v)
```

Comment:
There's no need to build the entire list. If it huges, it'll take hours for nothing.
You could use `first` which I like for this: https://pypi.org/project/first/
Otherwise:

```suggestion
        for interp in cls.find():
            if interp.matches(v):
                return interp
        raise FileNotFoundError("No interpreter matching %r", v)
```